### PR TITLE
add events support

### PR DESCRIPTION
**Pull request by comboy**
I wanted to submit events to datadog using elixir, so I added some basic support for them. 

Still learning the language, so I'm happy to improve the PR if there's something that can be done better.

### CODE REVIEW

File: lib/ex_statsd.ex [hunk 1]
```diff
@@ -271,6 +289,20 @@ defmodule ExStatsD do
     ] |> IO.iodata_to_binary
   end
 
+  defp packet({:event, title, text, opts}, _namespace, tags, _sample_rate) do
+    [
+      "_e",
+      "{#{title |> String.length},#{text |> String.length}}",
```
Comment (driv3r) [inline lib/ex_statsd.ex]: text needs to be escaped from new lines like `String.replace(text, "\n", "\\n")`

File: lib/ex_statsd.ex [hunk 2]
```diff
@@ -271,6 +289,20 @@ defmodule ExStatsD do
...
+      ":#{title}|#{text}",
+      opts[:priority]         && "|p:#{opts[:priority]}" || "",
+      opts[:alert_type]       && "|t:#{opts[:alert_type]}" || "",
+      opts[:source_type]      && "|s:#{opts[:source_type]}" || "",
+      opts[:aggregation_key]  && "|k:#{opts[:aggregation_key]}" || "",
+      opts[:hostname]         && "|h:#{opts[:hostname]}" || "",
+      tags |> tags_suffix
+    ] |> IO.iodata_to_binary
```
Comment (driv3r) [inline lib/ex_statsd.ex]: also it's probably also good to check the payload size like in ruby gem for example https://github.com/DataDog/dogstatsd-ruby/blob/master/lib/statsd.rb#L297:L318

**Comment by driv3r** [comment]
I would also add `date_happened` option, as it has higher priority from the one generated by datadog (it's when they receive the event)

File: lib/ex_statsd.ex (hunk 1 above)
Comment (comboy) [inline lib/ex_statsd.ex]: Why does it have to be? I tested some multiline events and it worked fine.

**Comment by comboy** [comment]
Sorry I don't know why I missed date_happened, I'll add that.

File: lib/ex_statsd.ex (hunk 2 above)
Comment (comboy) [inline lib/ex_statsd.ex]: Ah, good idea to check the lib maintained by them.  I didn't see anything about that in docs but if they have it in their library then I'll add it here too. Nice catch!

Comment (benwilson512) [inline lib/ex_statsd.ex]: Side note, I'm pretty sure we don't actually need to be doing the IO.iodata_to_binary function call, pretty sure :gen takes iodata.

**Comment by comboy** [comment]
* added <tt>date_happened</tt>
* replace from `\n` to `\\n` is indeed necessary, I don't know how I tested it earlier, fixed, thx @driv3r 
* kept original `source_type_name` as in documentation (maybe shorter would be nicer same with time instead of date_happened but I guess sticking to docs names makes it easier for people porting stuff from other languages)

And regarding size limit. I tested with some bigger message, it was cut off after 3996 characters, so I guess the msg size limit is 4KB. But.. I think raising as ruby library does is a bad idea. I don't think anybody want his instrumentation be the cause of errors (really hard to debug). So maybe just Logger.warn, but we would have to add Logger requirement (then the whole text could be put in a log message). For now I left it as it is, so only first ~4K characters will be available. Up to you.

**Comment by driv3r** [comment]
> replace from `\n` to `\\n` is indeed necessary, I don't know how I tested it earlier 
because it doesn't fail, DD just cuts the message after first `\n`

About the size, I think ~8KB is about whole payload, and not only the message, that's why it's been cut, maybe we could find something more in DD/statsd docs.

Throwing exceptions is also a bad idea, we could just truncate the text if necessary, or maybe DD do it already for us?

**Comment by comboy** [comment]
I did not find anything in docs, but if the message is big enough (tried something like 3MB) it doesn't get recorded at all, that's bad. So just truncation of text to 4KB (30KB still got sent but only ~4K chars visible on datadog) without any warning is OK? (plus some annotation in function doc)

And forgot to mention removed this iodata_to_binary as @benwilson512 mention, but needed to add it where sink is used

**Comment by comboy** [comment]
Added this limit and changed String.length to byte_size. I amended the last commit not to pollute history too much.

**Comment by benwilson512** [comment]
So, one overall question here. Are events datadog specific? Or is this a general statsd thing that datadog uses?

**Comment by driv3r** [comment]
It's datadog specific, just like tags

**Comment by bradcondo** [comment]
+1 Would like to see this get merged as well.

**Comment by KamilLelonek** [comment]
https://github.com/adamkittelson/dogstatsd-elixir supports it fortunately.

**Comment by bruce** [comment]
Changes pulled into #40.

**Comment by comboy** [comment]
"pulled" - so basically instead of doing a merge, you commit all these changes under your name, without any attribution. Not very encouraging for potential future contributors. But I'm glad it's in.

**Comment by bruce** [comment]
@comboy Despite the tone of your response, no slight was intended -- nor was committing "under my name" part of my reasoning (nor is it when I'm contributing to other open source projects). I apologize for the oversight; considering quite some time had passed since your PR was submitted, I didn't feel like it was fair to expect you to handle the merge, and I automatically squashed to keep the history as clean as possible -- we were hoping to do a quick (although very belated) release. I'm afraid in doing this, I ran afoul of GitHub's contribution graph rules -- which, as you obliquely point out, I suppose I don't pay enough attention to.

If you'd prefer, I'm happy to revert #40 and let you handle getting the PR into a mergeable state.

**Comment by bruce** [comment]
@comboy Alternately, if you have other ideas, I'm happy to entertain them.

**Comment by comboy** [comment]
@bruce personally I don't care that much but in general it irks me when contribution is not given in open source projects because for some people it makes difference to see their commits in the project or themselves on the github contributors list, and that's what drives their contributions.

I understand you didn't want to bother me with squash after a year, although they could have been squashed while keeping the author.

Anyway, no hard feelings from me and no change necessary here. Please just try to consider it in the future.

cheers

**Pull request closed without merging**